src/test/fuzz/fuzz.cpp

```diff
@@ -46,6 +46,9 @@ extern "C" int LLVMFuzzerInitialize(int* argc, char*** argv)
 
 // Declare main(...) "weak" to allow for libFuzzer linking. libFuzzer provides
 // the main(...) function.
+#if defined(MAC_OSX) && !defined(__AFL_COMPILER)
+extern int main(int argc, char** argv) __attribute__((weak_import));
```

Comment:
What is the point of declaring main here if it is already declared in libFuzzer?

I guess we could save us all of this `__attribute__` wrangling with a simple compile time check:

```cpp
#if defined(__AFL_COMPILER)
static bool read_stdin(std::vector<uint8_t>& data) { ... }
int main(int argc, char** argv) { ... }
#endif

Reply:
We could leave out the `main` declaration entirely, I just thought it might be clearer in regards to it being weak linked. If you prefer I can replace it with a comment as to why we special case libFuzzer on macOS.

Reply:
> I can replace it with a comment as to why we special case libFuzzer on macOS

My point is that we don't need to special case libFuzzer on macOS. I think, we can have the same simple code (a single compile time check) for all operating systems. There would be no need for `_attribute__` and `#if defined(MAC_OSX)` wrangling.

Reply:
Also, not having a main function should fix the issue where compilation and linking succeeds and the fuzzer just runs "blank", I think.

Reply:
>  should fix the issue where compilation and linking succeeds and the fuzzer just runs "blank", I think.

That is what is fixed by this PR in its current state. Are you still seeing that issue somewhere else with this change applied?

Reply:
Sorry, I think I might have been unclear. I see that your patch does fix the issue. However, I think we should go even further: A main function that does not exist should be better than a main function that is weak, as it would properly lead to linker issues instead of odd runtime behaviour.

The following patch (on top of master) should achieve the same result while also removing the main function altogether for fuzzers (except AFL).

```diff
diff --git a/src/test/fuzz/fuzz.cpp b/src/test/fuzz/fuzz.cpp
index a6ab620e21..5a7b96c89b 100644
--- a/src/test/fuzz/fuzz.cpp
+++ b/src/test/fuzz/fuzz.cpp
@@ -44,9 +44,9 @@ extern "C" int LLVMFuzzerInitialize(int* argc, char*** argv)
     return 0;
 }
 
-// Declare main(...) "weak" to allow for libFuzzer linking. libFuzzer provides
-// the main(...) function.
-__attribute__((weak)) int main(int argc, char** argv)
+// The fuzzer generally provides the main function, except for AFL
+#if defined(__AFL_COMPILER)
+int main(int argc, char** argv)
 {
     initialize();
 #ifdef __AFL_INIT
@@ -74,3 +74,4 @@ __attribute__((weak)) int main(int argc, char** argv)
 #endif
     return 0;
 }
+#endif

Reply:
Right, we can do that. I'll update this and drop the `__attribute__` usage in favour of an #ifdef.